## Part 1: The Doomsday Algorithm

The Doomsday algorithm, devised by mathematician J. H. Conway, computes the day of the week any given date fell on. The algorithm is designed to be simple enough to memorize and use for mental calculation.

__Example.__ With the algorithm, we can compute that July 4, 1776 (the day the United States declared independence from Great Britain) was a Thursday.

The algorithm is based on the fact that for any year, several dates always fall on the same day of the week, called the <em style="color:#F00">doomsday</em> for the year. These dates include 4/4, 6/6, 8/8, 10/10, and 12/12.

__Example.__ The doomsday for 2016 is Monday, so in 2016 the dates above all fell on Mondays. The doomsday for 2017 is Tuesday, so in 2017 the dates above will all fall on Tuesdays.

The doomsday algorithm has three major steps:

1. Compute the anchor day for the target century.
2. Compute the doomsday for the target year based on the anchor day.
3. Determine the day of week for the target date by counting the number of days to the nearest doomsday.

Each step is explained in detail below.

### The Anchor Day

The doomsday for the first year in a century is called the <em style="color:#F00">anchor day</em> for that century. The anchor day is needed to compute the doomsday for any other year in that century. The anchor day for a century $c$ can be computed with the formula:
$$
a = \bigl( 5 (c \bmod 4) + 2 \bigr) \bmod 7
$$
The result $a$ corresponds to a day of the week, starting with $0$ for Sunday and ending with $6$ for Saturday.

__Note.__ The modulo operation $(x \bmod y)$ finds the remainder after dividing $x$ by $y$. For instance, $12 \bmod 3 = 0$ since the remainder after dividing $12$ by $3$ is $0$. Similarly, $11 \bmod 7 = 4$, since the remainder after dividing $11$ by $7$ is $4$.

__Example.__ Suppose the target year is 1954, so the century is $c = 19$. Plugging this into the formula gives
$$a = \bigl( 5 (19 \bmod 4) + 2 \bigr) \bmod 7 = \bigl( 5(3) + 2 \bigr) \bmod 7 = 3.$$
In other words, the anchor day for 1900-1999 is Wednesday, which is also the doomsday for 1900.

__Exercise 1.1.__ Write a function that accepts a year as input and computes the anchor day for that year's century. The modulo operator `%` and functions in the `math` module may be useful. Document your function with a docstring and test your function for a few different years.  Do this in a new cell below this one.

In [1]:
import math
def anchorDay(year):
    """Return the anchor day for the year's century.
    
    Returns:
    ad: take values form integers 0 up to 6 corresponding to Sunday up to Saturday
    """
    
    y = int(str(year)[:2])
    ad = (5 * (y % 4) + 2) % 7
    return ad

### The Doomsday

Once the anchor day is known, let $y$ be the last two digits of the target year. Then the doomsday for the target year can be computed with the formula:
$$d = \left(y + \left\lfloor\frac{y}{4}\right\rfloor + a\right) \bmod 7$$
The result $d$ corresponds to a day of the week.

__Note.__ The floor operation $\lfloor x \rfloor$ rounds $x$ down to the nearest integer. For instance, $\lfloor 3.1 \rfloor = 3$ and $\lfloor 3.8 \rfloor = 3$.

__Example.__ Again suppose the target year is 1954. Then the anchor day is $a = 3$, and $y = 54$, so the formula gives
$$
d = \left(54 + \left\lfloor\frac{54}{4}\right\rfloor + 3\right) \bmod 7 = (54 + 13 + 3) \bmod 7 = 0.
$$
Thus the doomsday for 1954 is Sunday.

__Exercise 1.2.__ Write a function that accepts a year as input and computes the doomsday for that year. Your function may need to call the function you wrote in exercise 1.1. Make sure to document and test your function.

In [2]:
def doomsDay(year):
    """Return doomsday of the given year.
    
    Return:
    dd: take values from integers 0 up to 6 corresponding to Sunday up to Saturday
    """

    y = int(str(year)[-2:])
    dd = (y + math.floor(y)/4 + anchorDay(year)) % 7
    return int(dd)

### The Day of Week

The final step in the Doomsday algorithm is to count the number of days between the target date and a nearby doomsday, modulo 7. This gives the day of the week.

Every month has at least one doomsday:
* (regular years) 1/10, 2/28
* (leap years) 1/11, 2/29
* 3/21, 4/4, 5/9, 6/6, 7/11, 8/8, 9/5, 10/10, 11/7, 12/12

__Example.__ Suppose we want to find the day of the week for 7/21/1954. The doomsday for 1954 is Sunday, and a nearby doomsday is 7/11. There are 10 days in July between 7/11 and 7/21. Since $10 \bmod 7 = 3$, the date 7/21/1954 falls 3 days after a Sunday, on a Wednesday.

__Exercise 1.3.__ Write a function to determine the day of the week for a given day, month, and year. Be careful of leap years! Your function should return a string such as "Thursday" rather than a number. As usual, document and test your code.

In [3]:
def weekDay(month, day, year):
    """Return the day of week for a given date."""
    dds = {1:10, 2:28, 3:21, 4:4, 5:9, 6:6, 7:11, 8:8, 9:5, 10:10, 11:7, 12:12}
    wds = {0:'Sunday', 1:'Monday', 2:'Tuesday', 3:'Wednesday', 4:'Thursday', 5:'Friday', 6:'Saturday'}
    if year % 4 == 0:
        dds[1] += 1
        dds[2] += 1
    if (day - dds[month] + doomsDay(year)) % 7 >= 0:
        dow = (day - dds[month] + doomsDay(year))% 7
    else:
        dow = 7 - abs(day - dds[month] + doomsDay(year)) % 7
    return wds[dow]

__Exercise 1.4.__ How many times did Friday the 13th occur in the years 1900-1999? Does this number seem to be similar to other centuries?

In [4]:
def dayoccurance(weekday, day, year_start, year_end):
    """Return the number of occurance of a given weekday with the same day. """
    t = 0
    for year in xrange(year_start, year_end + 1):
        for month in xrange(1,13):
            if weekDay(month, day, year) == weekday:
                t += 1
    return t

In [5]:
print dayoccurance('Friday', 13, 1900, 1999)
print dayoccurance('Friday', 13, 1800, 1899)
print dayoccurance('Friday', 13, 1700, 1799)

173
172
173


There are 173 times that Friday the 13th occured. And this number seems to be similar to those (172 in 18th century and 173 in 17th century) of other centuries.

__Exercise 1.5.__ How many times did Friday the 13th occur between the year 2000 and today?

In [6]:
t1 = dayoccurance('Friday', 13, 2000, 2017)
t2 = weekDay(1, 13, 2017) == 'Friday'
t1 + t2

32

There are 32 times that Friday the 13th occur between the year 2000 and today(1/21/2017).

## Part 2: 1978 Birthdays

__Exercise 2.1.__ The file `birthdays.txt` contains the number of births in the United States for each day in 1978. Inspect the file to determine the format. Note that columns are separated by the tab character, which can be entered in Python as `\t`. Write a function that uses iterators and list comprehensions with the string methods `split()` and `strip()` to  convert each line of data to the list format

```Python
[month, day, year, count]
```
The elements of this list should be integers, not strings. The function `read_birthdays` provided below will help you load the file.

In [7]:
def read_birthdays(file_path):
    """Read the contents of the birthdays file into a string.
    
    Arguments:
        file_path (string): The path to the birthdays file.
        
    Returns:
        string: The contents of the birthdays file.
    """
    with open(file_path) as file:
        return file.read()

In [27]:
raw = read_birthdays('/Users/Rachel/Desktop/STA 141B/hw1/birthdays.txt')
start = raw.find('1/1/78')
end = raw.find('8028') + 4
bds = raw[start:end]
bds = bds.split('\n')
bds = [line.replace('78', '1978') for line in bds]
bds = [line.replace('\t', '/') for line in bds]
bds = [line.split('/') for line in bds]
birthdays = map(lambda line: [int(x) for x in line], bds)
print birthdays[:5]

[[1, 1, 1978, 7701], [1, 2, 1978, 7527], [1, 3, 1978, 8825], [1, 4, 1978, 8859], [1, 5, 1978, 9043]]


__Exercise 2.2.__ Which month had the most births in 1978? Which day of the week had the most births? Which day of the week had the fewest? What conclusions can you draw? You may find the `Counter` class in the `collections` module useful.

Create a new list of birthdays called birthdays_new including the day of week.

In [9]:
birthdays_new = birthdays
for line in birthdays_new:
    line.append(weekDay(line[0], line[1], line[2]))

In [30]:
# better solution
test = birthdays
test_new = [line + [weekDay(line[0], line[1], line[2])] for line in test]
# not map(lambda line: line.append(weekDay(line[0], line[1], line[2])), test)

Write a function that can return stats of birth, i.e births of every month, births of every day of week.

In [10]:
def bds_stats(stats_code):
    """Return a summary of births.
    
    Argument:
    stats_code(int): 0, 1, 4 reperesent 'month', 'day', 'day of week' respectively
    """
    bds_f = birthdays_new[:-1]
    bds_b = birthdays_new[1:]
    stats = {bds_f[0][stats_code]:bds_f[0][3]}
    for i,j in zip(bds_f, bds_b):
        if i[stats_code] == j[stats_code]:
            stats[i[stats_code]] += j[3]
        else:
            if j[stats_code] in stats:        # modified on 1/23/2017
                stats[j[stats_code]] += j[3]  # modified on 1/23/2017
            else:                             # modified on 1/23/2017
                stats[j[stats_code]] = j[3]   # modified on 1/23/2017
    return stats

May has the most births in 1978.

In [11]:
stats_m = bds_stats(0)
print stats_m
print max(stats_m, key = stats_m.get)

{1: 270695, 2: 629875, 3: 656584, 4: 1156577, 5: 1694812, 6: 270756, 7: 294701, 8: 1113795, 9: 293891, 10: 1500555, 11: 464671, 12: 474927}
5


Saturday has the most births and Thursday has the fewest births in 1978. 

In [12]:
stats_dow = bds_stats(4)
print stats_dow
print max(stats_dow, key = stats_dow.get)
print min(stats_dow, key = stats_dow.get)

{'Monday': 1389309, 'Tuesday': 1526458, 'Friday': 500541, 'Wednesday': 493897, 'Thursday': 493149, 'Sunday': 2083400, 'Saturday': 2335085}
Saturday
Thursday


__Exercise 2.3.__ What would be an effective way to present the information in exercise 2.2? You don't need to write any code for this exercise, just discuss what you would do.

A histogram or barplot of stats of births would be an informative presentation.